src/shared/lib/services/Logger.ts

```diff
@@ -0,0 +1,286 @@
+import { FileSystem } from 'react-native-file-access';
+import { FileLogger } from 'react-native-file-logger';
+import RNFetchBlob from 'rn-fetch-blob';
+
+import { FileService } from '@app/shared/api';
+
+import { IS_ANDROID } from '../constants';
+import { IMutex, Mutex, isAppOnline } from '../utils';
+
+const isAndroid = IS_ANDROID;
+
+const isIos = !isAndroid;
+
+interface ILogger {
+  log: (message: string) => void;
+  info: (message: string) => void;
+  warn: (message: string) => void;
+  error: (message: string) => void;
+  configure: () => void;
+  send: () => Promise<boolean>;
+  cancelSending: () => void;
+  clearAllLogFiles: () => Promise<void>;
+}
+
+type NamePath = {
+  fileName: string;
+  filePath: string;
+};
+
+type NamePathSize = {
+  size: number;
+} & NamePath;
+
+type FileExists = {
+  exists: boolean;
+} & NamePathSize;
+
+class Logger implements ILogger {
+  private mutex: IMutex;
+
+  private cancellationRequested: boolean;
+
+  constructor() {
+    this.mutex = Mutex();
+    this.cancellationRequested = false;
+  }
+
+  private async callWithMutex(func: () => void | Promise<any>) {
+    if (this.mutex.isBusy()) {
+      return;
+    }
+    try {
+      this.mutex.setBusy();
+      await func();
+    } finally {
+      this.mutex.release();
+    }
+  }
```

Comment:
Can we create a separate function decorator `withMutext`?
I would work like this:

Whenever you want to wrap a function with a mutex you do it this way:
`withMutext(mutex, () => somthing...)`

Or modify the Mutex function of even create a wrapper-class called CallableMutex which creates a mutex instance in its constructor and then reuses its methods. Then it would work like this:
```
class Logger {
...
  constructor() {
    this.mutex = CallableMutex();
    ...
  }
  ...  
  public log(message: string) {
    this.mutex.invoke(() => FileLogger.debug(message));
  }
}
```

Reply:
ok